File: packages/csharp/LinkedList.cs
namespace csharp_packages;

public class SinglyLinkedListNode<T> {
    public T Data;
    public SinglyLinkedListNode<T>? Next;
    public SinglyLinkedListNode(T data) { Data = data; }
}

public class SinglyLinkedList<T> {
    public SinglyLinkedListNode<T>? Head;
    public void Add(T data) {
        if (Head == null) { Head = new SinglyLinkedListNode<T>(data); return; }
        var curr = Head;
        while (curr.Next != null) curr = curr.Next;
        curr.Next = new SinglyLinkedListNode<T>(data);
    }
    
    public void Reverse() {
        SinglyLinkedListNode<T>? prev = null;
        SinglyLinkedListNode<T>? current = Head;
        SinglyLinkedListNode<T>? next = null;
        while (current != null) {
            next = current.Next;
            current.Next = prev;
            prev = current;
            current = next;
        }
        Head = prev;
    }
    
    public bool HasCycle() {
        if (Head == null) return false;
        SinglyLinkedListNode<T>? slow = Head;
        SinglyLinkedListNode<T>? fast = Head;
        while (fast != null && fast.Next != null) {
            slow = slow!.Next;
            fast = fast.Next.Next;
            if (ReferenceEquals(slow, fast)) return true;
        }
        return false;
    }
}

public class DoublyLinkedListNode<T> {
    public T Data;
    public DoublyLinkedListNode<T>? Next;
    public DoublyLinkedListNode<T>? Prev;
    public DoublyLinkedListNode(T data) { Data = data; }
}

public class DoublyLinkedList<T> {
    public DoublyLinkedListNode<T>? Head;
    public void Add(T data) {
        if (Head == null) { Head = new DoublyLinkedListNode<T>(data); return; }
        var curr = Head;
        while (curr.Next != null) curr = curr.Next;
        var node = new DoublyLinkedListNode<T>(data);
        curr.Next = node;
        node.Prev = curr;
    }
}

File: packages/zig/src/linkedlist.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn SinglyLinkedList(comptime T: type) type {
    return struct {
        pub const Node = struct {
            data: T,
            next: ?*Node = null,
        };

        head: ?*Node = null,
        allocator: Allocator,

        const Self = @This();

        pub fn init(allocator: Allocator) Self {
            return Self{ .allocator = allocator };
        }

        pub fn deinit(self: *Self) void {
            var curr = self.head;
            while (curr) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                curr = next;
            }
        }

        pub fn append(self: *Self, data: T) !void {
            const node = try self.allocator.create(Node);
            node.* = Node{ .data = data, .next = null };

            if (self.head == null) {
                self.head = node;
            } else {
                var curr = self.head;
                while (curr.?.next != null) {
                    curr = curr.?.next;
                }
                curr.?.next = node;
            }
        }

        pub fn reverse(self: *Self) void {
            var prev: ?*Node = null;
            var curr = self.head;
            var next: ?*Node = null;
            while (curr) |node| {
                next = node.next;
                node.next = prev;
                prev = node;
                curr = next;
            }
            self.head = prev;
        }

        pub fn hasCycle(self: *Self) bool {
            if (self.head == null) return false;
            var slow = self.head;
            var fast = self.head;
            while (fast != null and fast.?.next != null) {
                slow = slow.?.next;
                fast = fast.?.next.?.next;
                if (slow == fast) return true;
            }
            return false;
        }
    };
}

pub fn DoublyLinkedList(comptime T: type) type {
    return struct {
        pub const Node = struct {
            data: T,
            next: ?*Node = null,
            prev: ?*Node = null,
        };

        head: ?*Node = null,
        tail: ?*Node = null,
        allocator: Allocator,

        const Self = @This();

        pub fn init(allocator: Allocator) Self {
            return Self{ .allocator = allocator };
        }

        pub fn deinit(self: *Self) void {
            var curr = self.head;
            while (curr) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                curr = next;
            }
        }

        pub fn append(self: *Self, data: T) !void {
            const node = try self.allocator.create(Node);
            node.* = Node{ .data = data, .next = null, .prev = self.tail };

            if (self.tail) |tail| {
                tail.next = node;
            } else {
                self.head = node;
            }
            self.tail = node;
        }
    };
}